_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+            i < UncheckedCounter.wrap(20);
+            i = i + ONE
+        )
+            ++internalCounter;
+
+        return internalCounter;
+    }
+}
+```
+
+### AST Changes
+The new feature introduces two new attributes to the Abstract Syntax Tree (AST) produced by the compiler:
+1. `functionList` on `UsingForDirective` nodes can now contain operator entries of the following structure:
+    ```json
+    {
+        "operator": "+",
+        "definition": { /* function */ }
+    }
+    ```
+    in addition to the current entries for functions that are structured as follows:
+    ```json
+    {
+        "function": { /* function */ }
+    }
+    ```
+1. `BinaryOperation` and `UnaryOperation` nodes may now have a `function` attribute.
+    If present, the attribute indicates that the operation is actually an invocation of a user-defined
+    operator and specifies the ID of the function used to define it.
+
+This effectively introduces a new way to perform a function call, not represented by a `FunctionCall` node.
+This will affect any tool that needs to detect and follow function calls, for example, to build a
+control flow graph.
+
+### Design Rationale
+
+#### Why Does Solidity Need User-Defined Operators?
+The support for user-defined value types in Solidity cannot be considered complete without more natural
+ways to operate on them.
+The ability to use operators was envisioned from the beginning, and the main reason it was not provided
+immediately was the uncertainty regarding interactions with other features we plan to introduce.
+We decided that it was best to release the feature in a minimal usable state and extend it over time,
+taking into account user feedback.
+
+UDVTs in general were meant to fulfill two needs:
+1. More type-safety when performing operations on values of the same underlying value type that express
+    conceptually different things.
+    For example quantities of different tokens could both be stored as `uint` but adding them without
+    an explicit conversion is a mistake that can only be prevented if the compiler knows that they
+    are values of different kinds.
+
+1. The ability to define new types using value types as the underlying representation.
+    The best example of this are fixed-point types.
+
+    Fixed-point types were originally going to be an integral part of the language, and the
+    partially implemented [`ufixed` and `fixed`](https://docs.soliditylang.org/en/latest/types.html#fixed-point-numbers)
+    types are the vestiges of that old design.
+    While trying to agree on the final design for the feature, however, we realized that no implementation
+    is going to be a trade-off good enough for most situations and cannot become a suitable default choice.
+    We decided to leave the choice up to library authors and instead focus on providing primitives
+    necessary to build their own fixed-point types that feel like a native type.
```

Comment:
```suggestion
    While trying to agree on the final design for the feature, however, we realized that no implementation
    will be a good enough trade-off for most situations and cannot become a suitable choice by default.
    We decided to leave the choice up to library authors and instead focus on providing primitives
    necessary to build their own fixed-point types that feel like native types.
```

Reply:
Changed, except `suitable choice by default`, which is not quite what I meant. I did mean the "default choice", i.e. a fixed point implementation that is the default one in the language.